### ISSUE TITLE
X710 NIC running problems

### ISSUE THREAD

**Issue by killerstemp** [eunyoung14/mtcp#122]
X710 NIC RSS algorithm and 82599 is not the same, resulting in mtcp server load balance andf the network card RSS Inconsistent

**Comment by ajamshed** [comment]
Hi Killerstemp,

Apologies for the delayed response. I will be committing some new code shortly. This shall enable support for Intel X710 chipsets. Thereafter, mTCP-based server applications should start working. I will let you know once the code is checked in.

You are right. The current stack does not support ``mtcp_init_rss()`` functionality for X710 adapters. I still need to reverse engineer the RSS computation for these types of NICs.

**Comment by ajamshed** [comment]
I have added code in the development branch of mTCP. Server-side applications will run fine with this patch. I will check in more changes tomorrow. I figured out a way to get ``mtcp_init_rss()`` working as well. I will clean up the code and merge the changes to the master branch within a couple of days.

**Comment by ajamshed** [comment]
Let me know if you face any issues. This X710 patch has only been tested only by me :).

**Comment by ajamshed** [comment]
FYI: I have added code that would get XL710 network adapters compatible with mTCP.

**Comment by killerstemp** [comment]
Very grateful to your efforts! Mtcp current data plan and socket are not separated are running on the same core, will cause high load when the packet loss to reduce performance, there is no good solution :)

**Comment by ajamshed** [comment]
I'm not sure whether I understood your comment. Just FYI, these are the performance numbers that I am getting:

Testbed:
Server/Client machines (both have the same CPU)
Intel(R) Xeon(R) CPU E5-2690 0 @ 2.90GHz (2 x 8 cores)
20 MB L3 cache size
132 GB RAM

Testsuite:
``lighttpd-1.4.32``
``epwget`` (16,000 concurrent connections)
``Linux-4.4``

File size [B] | Throughput [Gbps]
64 | 10.0
128 | 10.7
256 | 12.2
512 | 15.5
1K | 21.7
2K | 20.7
4K | 25.4
8K | 26.3
16K | 26.8


The maximum theoretical bandwidth is bound by the PCIv3 (8-lane) capacity (~56 Gbps). So I know that mTCP is off by 10s of Gbps. For previous cases (10 Gbps adapters), the link was always the bottleneck. So we are now facing some new (previously unknown) bottlenecks in the code. I profiled the code. I think I can improve the performance by at least 5 Gbps easily. I will commit some changes to the code in the coming days. The number should touch ~30 Gbps in the best case. Then I will again profile the code more :). I will let you know once I have updated the code.  

FYI: linux-based TCP stack is performing much poorly (~14.0 Gbps in the best case).
